### ISSUE TITLE
Type parameters with constraints should be reconstructable

### ISSUE THREAD

**Issue by zach-waggoner** [microsoft/TypeScript#31994]
**TypeScript Version:**  3.6.0-dev.20190619

**Search Terms:** type parameter generic constraint assignable instantiated subtype reconstruct

**Code**

```ts
interface AConstraint {
  x: boolean;
}

function fn<A extends AConstraint>(arg: A): A {
  const { x, ...rest } = arg;
  return { x, ...rest };
}
```

**Expected behavior:**

pass

**Actual behavior:**

```
test.ts:7:3 - error TS2322: Type '{ x: boolean; } & Pick<A, Exclude<keyof A, "x">>' is not assignable to type 'A'.
  '{ x: boolean; } & Pick<A, Exclude<keyof A, "x">>' is assignable to the constraint of type 'A', but 'A' could be instantiated with a different subtype of constraint 'AConstraint'.
```

It seems like this should work, since `{ x, ...rest }` should be assignable to `A` regardless of how `A` is instantiated.

On a related note, I've found that the following does not throw an error:

```ts
interface AConstraint {
  x: boolean;
  y: boolean;
}

function fn <A extends { x: boolean; y: boolean }> (arg: A): A['x'] {
  return arg.y;
}
```

This also seems like a bug, since `A['x']` may not be the same type as `A['y']` after `A` is instantiated (e.g., if `A` is instantiated as `{ x: true; y: false }`).

**Playground Link:** 

https://www.typescriptlang.org/play/#src=interface%20AConstraint%20%7B%0D%0A%20%20x%3A%20boolean%3B%0D%0A%7D%0D%0A%0D%0Afunction%20fn%3CA%20extends%20AConstraint%3E(arg%3A%20A)%3A%20A%20%7B%0D%0A%20%20const%20%7B%20x%2C%20...rest%20%7D%20%3D%20arg%3B%0D%0A%20%20return%20%7B%20x%2C%20...rest%20%7D%3B%0D%0A%7D%0D%0A

**Related Issues:** 

https://github.com/Microsoft/TypeScript/issues/29049

**Comment by fatcerberus** [comment]
The error seems correct: you have a constraint on `A` that it's at least `AConstraint` so you can work with the value you have as though it's an `AConstraint`, but you can't actually *produce* a value of type `A` because you don't know how specific `A` actually is.  There's no lower bound, so it could be any subtype at all.

For example:
```ts
function fn<A extends Animal>(arg: A): A
{
    return new Giraffe("Jeff");  // Giraffe is assignable to Animal, but...
}

let cat = new Cat("Mittens");
let newCat = fn(cat);  // ...we just wanted a cat. :-(
```

As you point out, in your case you capture `...rest` which should cover all the bases in practice--*but* I think TS is just assuming the worst: `A` might be `never`.  Then nothing at all would be assignable to it.

**Comment by zach-waggoner** [comment]
That could be. After further digging, I suspect this might be related to #28884. I will leave this open for now in case my second code example is a distinct issue.
